src/Qir/Runtime/lib/QIR/QubitManagerRestrictedReuse.hpp

```diff
@@ -0,0 +1,210 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include <csignal>
+#include <iostream>
+#include <vector>
+
+#include "CoreTypes.hpp"
+
+namespace Microsoft
+{
+namespace Quantum
+{
+
+    // TODO: Remove this! Use standard ones.
+    static void FailNow(const char* message) { std::cout << message; throw message; }
+    static void FailIf(bool condition, const char* message) {
+        if (condition) {
+            FailNow(message);
+        }
+    }
+
+
+    // TODO: Decide on the interface and move this to appropriate place.
+    struct IQubitManager
+    {
+        virtual ~IQubitManager() = default;
+        virtual Qubit Allocate() = 0;
+        virtual void Release(Qubit qubit) = 0;
+    };
+
+    // The end of free lists are marked with NoneMarker value. It is used like null for pointers.
+    // This value is non-negative just like other values in the free lists.
+    constexpr int NoneMarker = INT_MAX;
+
+    // Explicitly allocated qubits are marked with Allocated value.
+    // Qubits implicitly allocated for borrowing are "refcounted" with values [Allocated+1 .. -2].
+    // When refcount needs to be decreased to Allocated value, the qubit is automatically released.
+    constexpr int AllocatedMarker = INT_MAX;
```

Comment:
FYI: The [`std::numeric_limits<int>::max()`](https://en.cppreference.com/w/cpp/types/numeric_limits) is considered the C++ counterpart of C's `INT_MAX`.

Reply:
Thanks! Updated.